From 09b9b54c88cb45e5892ca2b5e0f4e4e33877cc4a Mon Sep 17 00:00:00 2001
From: nyanmisaka <nst799610810@gmail.com>
Date: Sat, 1 Feb 2025 15:19:30 +0800
Subject: [PATCH] dict: Fallback to internal entropy

Signed-off-by: nyanmisaka <nst799610810@gmail.com>
---
 dict.c | 32 +++++++++++++++++++-------------
 1 file changed, 19 insertions(+), 13 deletions(-)

diff --git a/dict.c b/dict.c
index 49e1c6bf..301ef61b 100644
--- a/dict.c
+++ b/dict.c
@@ -962,9 +962,10 @@ xmlInitRandom(void) {
                                  BCRYPT_USE_SYSTEM_PREFERRED_RNG);
         if (!BCRYPT_SUCCESS(status)) {
             fprintf(stderr, "libxml2: BCryptGenRandom failed with "
-                    "error code %lu\n", GetLastError());
-            abort();
+                    "error code %lu, using internal entropy\n", GetLastError());
+            goto internal_entropy;
         }
+        return;
 #elif defined(HAVE_GETENTROPY)
         while (1) {
             if (getentropy(globalRngState, sizeof(globalRngState)) == 0)
@@ -972,20 +973,25 @@ xmlInitRandom(void) {
 
             if (errno != EINTR) {
                 fprintf(stderr, "libxml2: getentropy failed with "
-                        "error code %d\n", errno);
-                abort();
+                        "error code %d, using internal entropy\n", errno);
+                goto internal_entropy;
             }
         }
-#else
-        int var;
-
-        globalRngState[0] =
-                (unsigned) time(NULL) ^
-                HASH_ROL((unsigned) ((size_t) &xmlInitRandom & 0xFFFFFFFF), 8);
-        globalRngState[1] =
-                HASH_ROL((unsigned) ((size_t) &xmlRngMutex & 0xFFFFFFFF), 16) ^
-                HASH_ROL((unsigned) ((size_t) &var & 0xFFFFFFFF), 24);
+        return;
 #endif
+internal_entropy:
+        {
+            int var;
+
+            memset(globalRngState, 0, sizeof(globalRngState));
+
+            globalRngState[0] =
+                    (unsigned) time(NULL) ^
+                    HASH_ROL((unsigned) ((size_t) &xmlInitRandom & 0xFFFFFFFF), 8);
+            globalRngState[1] =
+                    HASH_ROL((unsigned) ((size_t) &xmlRngMutex & 0xFFFFFFFF), 16) ^
+                    HASH_ROL((unsigned) ((size_t) &var & 0xFFFFFFFF), 24);
+        }
     }
 }
 
-- 
2.34.1

