diff --git a/libavcodec/videotoolbox.c b/libavcodec/videotoolbox.c
index c234b18f59..72e8137b35 100644
--- a/libavcodec/videotoolbox.c
+++ b/libavcodec/videotoolbox.c
@@ -983,6 +983,14 @@ static int videotoolbox_start(AVCodecContext *avctx)
             av_log(avctx, AV_LOG_ERROR, "VideoToolbox reported invalid data.\n");
             return AVERROR_INVALIDDATA;
         case 0:
+            if (avctx->skip_frame >= AVDISCARD_NONKEY) {
+                status = VTSessionSetProperty(videotoolbox->session,
+                                              kVTDecompressionPropertyKey_OnlyTheseFrames,
+                                              kVTDecompressionProperty_OnlyTheseFrames_KeyFrames);
+                if (status) {
+                    av_log(avctx, AV_LOG_WARNING, "kVTDecompressionProperty_OnlyTheseFrames_KeyFrames is not supported on this device. Ignoring.\n");
+                }
+            }
             return 0;
         default:
             av_log(avctx, AV_LOG_ERROR, "Unknown VideoToolbox session creation error %d\n", (int)status);
