From 5e8e2f3861dfbc4864490401c5209fbc24df3c29 Mon Sep 17 00:00:00 2001
From: Reimar Döffinger <Reimar.Doeffinger@gmx.de>
Date: Wed, 12 Nov 2014 09:53:01 +0100
Subject: [PATCH] configure: Hack to treat x32 as x86_64.

---
 configure | 3 +++
 1 file changed, 3 insertions(+), 0 deletions(-)

diff --git a/configure b/configure
index a2d872b..8ad4777 100755
--- a/configure
+++ b/configure
@@ -3933,6 +3933,9 @@ case "$arch" in
     ;;
     x86)
         check_64bit x86_32 x86_64 'sizeof(void *) > 4'
+        # Treat x32 as x64 for now. Note it also needs spic=$shared
+        test "$subarch" = "x86_32" && check_cpp_condition stddef.h 'defined(__x86_64__)' &&
+            subarch=x86_64
         if test "$subarch" = "x86_64"; then
             spic=$shared
         fi
--
libgit2 0.21.0

