From: Carl Eugen Hoyos <cehoyos@ag.or.at>
Date: Tue, 6 Sep 2016 12:47:34 +0200
Subject: ffmpeg_opt: Suggest to use "file:..." if a protocol was not found.

Fixes Debian bug 785690.
---
 ffmpeg_opt.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/ffmpeg_opt.c b/ffmpeg_opt.c
index 7785a30..f96c655 100644
--- a/ffmpeg_opt.c
+++ b/ffmpeg_opt.c
@@ -982,6 +982,8 @@ static int open_input_file(OptionsContext *o, const char *filename)
     err = avformat_open_input(&ic, filename, file_iformat, &o->g->format_opts);
     if (err < 0) {
         print_error(filename, err);
+        if (err == AVERROR_PROTOCOL_NOT_FOUND)
+            av_log(NULL, AV_LOG_ERROR, "Did you mean file:%s?\n", filename);
         exit_program(1);
     }
     if (scan_all_pmts_set)
